core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -683,8 +704,7 @@ private[deploy] class Master(
       if (app.coresLeft >= coresPerExecutor) {
         // Filter out workers that don't have enough resources to launch an executor
         val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
-          .filter(worker => worker.memoryFree >= app.desc.memoryPerExecutorMB &&
-            worker.coresFree >= coresPerExecutor)
+          .filter(canLaunchExecutor(_, app.desc))
           .sortBy(_.coresFree).reverse
         val assignedCores = scheduleExecutorsOnWorkers(app, usableWorkers, spreadOutApps)
```

Comment:
`scheduleExecutorsOnWorkers()` currently has two modes: 1) one executor per worker 2) multiple executors per worker.  If we don't configured `coresPerExecutor`, then, we use mode 1. In mode 1, only one executor can be launched per worker and that executor grabs all cores on the worker. But, for memory and resources, it still respects to memoryPerExecutor and resourcesPerExecutor rather than requires all.